# Proyecto Modulo 1: Optimización en Dosificación del  Concreto

![ITESO](https://st2.depositphotos.com/1111050/11917/i/950/depositphotos_119178666-stock-photo-a-concrete-building.jpg)

> ### Antonio Morales  
> ### Fernando Jiménez  
> ### Lorena Villafaña

### Objetivos generales

- Lograr planetear un problema de optimización resultado de un area de nuestra carrera (Ingenieria Civil).
- Comprehender lo planteado y los resultados optenidos por esta optimización para su aplicacion.

### Objetivos especificos

- Plantear las restricciones de manera analoga para posteriormente plantearlas en matrices y en codigo phytom. 
- Resolver el problema de optimización usando linprog.
- Una vez con los resultados, usar estos para resolver el problema inicial y conocer un precio minimo. 

Uno de los problemas que se presentan en obra es la preparación del concreto segun la resistencia necesaria para el proyecto. Porque aunque la dosificación esta establecida, en la práctica no se repesta totalmete por diferentes factores.
En este proyecto queremos asegurar que la preparación de concreto sea viable en obra, apesar de las variaciones en la dosificacion y siga cumpliendo con la resistencia adecuada( **F'c=250 Kg/cm2**)pero con un costo **Mínimo**.

### Minimizar el costo

El análisis de las cantidades necesarias se hará en base a las especificaciones para el concreto  F'c=250 KG/CM2.  
Nuestras variables son los compuestos a usar en la mezcla, para simplificar su notación:
+ $x_1:$ Cemento 
+ $x_2:$ Grava  
+ $x_3:$ Arena  
+ $x_4:$ Agua  


En la siguiente tabla se establecen los datos _(en porcentajes)_ de la cantidad permitida de variación en cada compuesto en el **total del mezcla**:


Compuesto|Porcentaje Permitido(%)
:----|----
Cemento|  16 - 18
Grava|  37 - 47
Arena|  24 - 34
Agua|  11 - 15


La dosificación está en **"Litros"**, ya que es la manera de medicicón usada en obra. Los costos de cada elemento son:

Ingrediente|Costo por litro
:----|----
Cemento|$\$$4.39
Grava|$\$$3.70
Arena|$\$$2.70
Agua|$\$$1.50
  

###  Funcion objetivo
Toma los valores del costo por litro de cada compuesto, ya que el objetivo es minimizar el costo.
+ $4.39x_1$ + $3.70x_2 +2.70x_3 + 1.50x_4 $ = minimizar costo


### Restricciones 
**Restriciones de desigualdad**

+ $x_1\geq 16$
+ $x_1\leq 18$

+ $x_2\geq 37$
+ $x_2\leq 47$

+ $x_3\geq 24$
+ $x_3\leq 34$

+ $x_4\geq 11$
+ $x_4\leq 15$

+ $x_2+x_3\geq70$
+ $x_2+x_3\leq75$

**Restricciones de igualdad**:el porcentaje total(sumado de cada elemento) debe ser igual a 100%, al mencionar que es el total de la mezcla

+ $x_1+x_2+x_3+x_4 = 100$

Las restricciones que tienen un mayor que son multiplicadas por **-1** ,para cambiar de sentido del operador 

Las operaciones se harán con la siguientes matrices y vectores.
- $\boldsymbol{c}=\left[4.39\quad 3.70\quad 2.70\quad 1.50\right]^\top$,
- $\boldsymbol{A_ub}=\left[\begin{array}{cc}-1 & 0 & 0 & 0 
\\ 1 & 0 & 0 & 0 
\\ 0 & -1 & 0 & 0 
\\ 0 & 1 & 0 & 0
\\ 0 & 0 & -1 & 0 
\\ 0 & 1 & 1 & 0 
\\ 0 & 0 & 0 & -1
\\ 0 & 0 & 0 & 1 
\\ 0 & -1 & -1 & 0 
\\ 0 & 1 & 1 & 0 
\end{array}\right]$, 
- $\boldsymbol{b_ub}=\left[-16\quad 18\quad -37\quad 47\quad -24\quad 34\quad -11\quad 15\quad -70\quad 75\right ]^\top$.
- $\boldsymbol{A_eq}=\left[\begin{array}{cc}1 & 1 & 1 & 1 \end{array}\right]$,
- $\boldsymbol{b_eq}=\left[100\right]^\top$.

  

El procedimiento para encontrar los valores es el sigueinte:

In [1]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import sympy as sym
import scipy.optimize as opt

In [2]:

c = np.array([4.39,3.70,2.70,1.50])

A_ub=np.array([[-1,0,0,0],[1,0,0,0],[0,-1,0,0],[0,1,0,0],[0,0,-1,0],[0,0,1,0],[0,0,0,-1],[0,0,0,1],[0,-1,-1,0],[0,1,1,0]])
b_ub=np.array([-16,18,-37,47,-24,34,-11,15,-70,75])

A_eq=np.array([[1,1,1,1]])# restricciones de igualdad

b_eq=np.array([100])

In [3]:
resultado=opt.linprog(c,A_ub,b_ub,A_eq,b_eq)
resultado

     con: array([0.])
     fun: 317.24
 message: 'Optimization terminated successfully.'
     nit: 9
   slack: array([ 0.,  2.,  0., 10.,  9.,  1.,  3.,  1.,  0.,  5.])
  status: 0
 success: True
       x: array([16., 37., 33., 14.])

El resultado para que el sea minimo el gasto en materiales y se cumplan las restricciones son las siguientes:
 
- Usar **16%** de cemento.
- Usar **37%** de grava.
- Usar **33%** de arena.
- Usar **14%** de agua.




**Problema Aplicado (trabe)**

Se tiene una trabe con las siguientes dimensiones: 

- base=40cm

- peralte=1m

- claro=15m

Calcule cuantos metros cubicos de cemento / grava / arena / agua se requieren para sasisfacer la demanda de materiales segun las restricciones y calcule el costo.

El volumen d la trabe esta dado por la formula:

$$V=bhl$$

Aplicando la formula, tenemos que:

$$V=(0.40m)(1.00m)(15m)=6m^3$$

Calculamos los porcentajes de los materiales:

$$Cemento=(6m^3)(0.16)=0.96m^3$$
$$Grava=(6m^3)(0.37)=2.22m^3$$
$$Arena=(6m^3)(0.33)=1.98m^3$$
$$Agua=(6m^3)(0.14)=0.84m^3$$

Calculamos el costo total de los materiales segun la tabla de costos:

$$Cemento=($4.39)(960lt)=$4214$$
$$Grava=($3.70)(222lt)=$8214$$
$$Arena=($2.70)(1980lt)=$5346$$
$$Agua=($1.50)(840lt)=$1260$$

Resultados finales:

**Costo final de la trabe 19,034 mnx**


## **Conclusion:**

Para concluir este trabajo se puede decir que este problema de optimización si fue realizado exitosamente, ya que los resultados concuerdan con todas las restricciones minimizando el costo total.